### ISSUE TITLE
Fetching should be defined in terms of Fetch

### ISSUE THREAD

**Issue by annevk** [fedidcg/FedCM#261]
Requirements such as

> This request MUST NOT follow [HTTP redirects](https://httpwg.org/specs/rfc7231.html#header.location) and instead abort with an error if there are any.

are not needed if this specification would build on top of Fetch. That would also address a number of other things that are unclear, such as referrer handling, etc.

**Comment by cbiesinger** [comment]
duplicate of issue 188. (what is unclear about referer handling?)

**Comment by annevk** [comment]
How is it clear? :-)

**Comment by cbiesinger** [comment]
Heh. Well each endpoint specifies whether it is sent or not, e.g.:
https://fedidcg.github.io/FedCM/#idp-api-client-id-metadata-endpoint
> c) with a [Referer](https://httpwg.org/specs/rfc7231.html#header.referer) header indicating the [RP](https://fedidcg.github.io/FedCM/#rp)'s origin (as if [Referer-Policy: strict-origin](http://www.w3.org/TR/referrer-policy/#referrer-policy-strict-origin) was in use)

I don't disagree with the fetch suggestion, but is that line not clear?

**Comment by cbiesinger** [comment]
Ah yes. That should be better integrated with the other section, thanks for pointing that out

**Comment by cbiesinger** [comment]
I'm going to reopen this issue and narrow the scope to clarifying that issue

@npm1

**Comment by cbiesinger** [comment]
(I can't figure out how to change the title of a github issue...)

**Comment by annevk** [comment]
(next to the green "New issue" button there's a gray "Edit" button, both to the right of the heading)

**Comment by cbiesinger** [comment]
I must be missing some permission to do that :(

**Comment by npm1** [comment]
It may be only @samuelgoto has the permissions to edit...

**Comment by npm1** [comment]
We integrated one of the fetch calls in here https://fedidcg.github.io/FedCM/#check-the-root-manifest. Does this look OK? If so then we can move ahead and integrate the others similarly, if not let's improve the integration of that one more and then model the other ones on top of this one.

**Comment by annevk** [comment]
> Set rootUrl’s [path](https://url.spec.whatwg.org/#concept-url-path) to ".well-known/web-identity".

This needs to be a list, no? You might want to assert provider’s configURL's scheme for clarity here.

> If response’s [Content-Type Metadata](https://html.spec.whatwg.org/multipage/urls-and-fetching.html#content-type) is not a [JSON MIME Type](https://mimesniff.spec.whatwg.org/#json-mime-type), return false.

You want to use https://fetch.spec.whatwg.org/#concept-header-extract-mime-type here.


You also need to set many more fields of the request, as discussed in some of the other issues.

**Comment by annevk** [comment]
Also, what you have attempted to define is that "check the root manifest" returns a boolean but instead it returns early and then the callback passed to processResponseConsumeBody returns a boolean which is then ignored.

**Comment by cbiesinger** [comment]
for reference, https://github.com/whatwg/fetch/pull/1495 (if accepted) spec's a "web-identity" value for Sec-Fetch-Dest

**Comment by npm1** [comment]
Closing but feel free to provide feedback if there's something off with how we integrated with Fetch.

**Comment by annevk** [comment]
"check the root manifest" is still broken as it expects fetch to block or some such.

**Comment by npm1** [comment]
@annevk help me understand how to actually use the `fetch` result? We have an algorithm `alg` that wants to call `fetch` and return something based on `processResponseConsumeBody`. Currently we have
```
alg:
...
1. Create |var|
2. Fetch and in processResponseConsumeBody set |var| to something
3. Return |var|
```

Is there a notion of 'await' in the spec so this becomes correct? Would it be ok to have step 3 be `Wait until |var| is no longer null` or some such? Otherwise using `fetch` on algorithms that return something based on it seems really complicated. Any example of correct usage? I was browsing the HTML spec and most fetch usages are not in returning algorithms. One that did is [broken in the same way](https://html.spec.whatwg.org/#poster-frame).

**Comment by annevk** [comment]
I mean in general you cannot really invoke fetch from "in parallel" without a lot of work so you'll have to account for it "calling you back" at some point. Otherwise you'd have to snapshot a bunch of state from the global object and take that with you somehow (because Fetch will need it for enforcing CSP and other things). And none of that is really worked out in detail. (And yeah, some of HTML is indeed broken in this way, but hopefully for new specs we don't repeat those mistakes.)

**Comment by domfarolino** [comment]
Is this done now that #378 has landed?
